--- Makefile.orig	2012-03-24 14:43:05.000000000 -0700
+++ Makefile	2012-03-24 15:35:22.000000000 -0700
@@ -47,6 +47,7 @@
 		pfx=lib; \
 		if [ "$(PLATFORM)" != "Cygwin" ]; then \
 			case "$(CFLAGS)" in \
+			*OPENSSL_NO_HW*)	echo ... skipping \(OPENSSL_NO_HW\); exit 0;; \
 			*DSO_BEOS*) sfx=".so";; \
 			*DSO_DLFCN*) sfx=`expr "$(SHLIB_EXT)" : '.*\(\.[a-z][a-z]*\)' \| ".so"`;; \
 			*DSO_DL*) sfx=".sl";; \
