--- rbconfig.rb.orig	2012-03-25 14:00:32.000000000 -0700
+++ rbconfig.rb	2012-03-25 14:10:52.000000000 -0700
@@ -3,8 +3,8 @@
 # changes made to this file will be lost the next time ruby is built.
 
 module RbConfig
-  RUBY_VERSION == "1.9.2" or
-    raise "ruby lib version (1.9.2) doesn't match executable version (#{RUBY_VERSION})"
+#  RUBY_VERSION == "1.9.2" or
+#    raise "ruby lib version (1.9.2) doesn't match executable version (#{RUBY_VERSION})"
 
   TOPDIR = File.dirname(__FILE__).chomp!("/lib/ruby/1.9.1/x86_64-mingw32")
   DESTDIR = '' unless defined? DESTDIR
