--- ./chardefs.h.orig	2012-03-22 17:33:17.000000000 -0700
+++ ./chardefs.h	2012-03-22 17:58:29.000000000 -0700
@@ -53,6 +53,11 @@
 /* Some character stuff. */
 #define control_character_threshold 0x020   /* Smaller than this is control. */
 #define control_character_mask 0x1f	    /* 0x20 - 1 */
+#define control_character_bit 0x40	    /* 0x000000, must be off. */
+
+/* Some character stuff. */
+#define control_character_threshold 0x020   /* Smaller than this is control. */
+#define control_character_mask 0x1f         /* 0x20 - 1 */
 #define meta_character_threshold 0x07f	    /* Larger than this is Meta. */
 #define control_character_bit 0x40	    /* 0x000000, must be off. */
 #define meta_character_bit 0x080	    /* x0000000, must be on. */
