--- ./rlshell.h.orig	2012-03-22 17:33:17.000000000 -0700
+++ ./rlshell.h	2012-03-22 17:58:29.000000000 -0700
@@ -25,10 +25,10 @@
 
 #include "rlstdc.h"
 
-extern char *sh_single_quote PARAMS((char *));
-extern void sh_set_lines_and_columns PARAMS((int, int));
-extern char *sh_get_env_value PARAMS((const char *));
-extern char *sh_get_home_dir PARAMS((void));
-extern int sh_unset_nodelay_mode PARAMS((int));
+READLINE_DLL_IMPEXP char *sh_single_quote PARAMS((char *));
+READLINE_DLL_IMPEXP void sh_set_lines_and_columns PARAMS((int, int));
+READLINE_DLL_IMPEXP char *sh_get_env_value PARAMS((const char *));
+READLINE_DLL_IMPEXP char *sh_get_home_dir PARAMS((void));
+READLINE_DLL_IMPEXP int sh_unset_nodelay_mode PARAMS((int));
 
 #endif /* _RL_SHELL_H_ */
