--- ./rlstdc.h.orig	2012-03-22 17:33:17.000000000 -0700
+++ ./rlstdc.h	2012-03-22 17:58:29.000000000 -0700
@@ -42,4 +42,28 @@
 #  endif
 #endif
 
+#ifndef __GNUC__
+# define __DLL_IMPORT__  __declspec(dllimport)
+# define __DLL_EXPORT__  __declspec(dllexport)
+#else
+# define __DLL_IMPORT__  __attribute__((dllimport)) extern
+# define __DLL_EXPORT__  __attribute__((dllexport)) extern
+#endif 
+
+#if (defined __WIN32__) || (defined _WIN32)
+# ifdef BUILD_READLINE_DLL
+#  define READLINE_DLL_IMPEXP     __DLL_EXPORT__
+# elif defined(READLINE_STATIC)
+#  define READLINE_DLL_IMPEXP     extern
+# elif defined (USE_READLINE_DLL)
+#  define READLINE_DLL_IMPEXP     __DLL_IMPORT__
+# elif defined (USE_READLINE_STATIC)
+#  define READLINE_DLL_IMPEXP     extern
+# else /* assume USE_READLINE_DLL */
+#  define READLINE_DLL_IMPEXP     __DLL_IMPORT__
+# endif
+#else /* __WIN32__ */
+# define READLINE_DLL_IMPEXP
+#endif
+
 #endif /* !_RL_STDC_H_ */
