--- include/libxml/xmlexports.h.orig	2012-03-25 10:31:23.000000000 -0700
+++ include/libxml/xmlexports.h	2012-03-25 10:40:33.000000000 -0700
@@ -113,7 +113,8 @@
    * _imp__xmlFree listed as missing. Try to workaround the problem
    * by also making that declaration when compiling client code.
    */
-  #if defined(IN_LIBXML) && !defined(LIBXML_STATIC)
+  /* #if defined(IN_LIBXML) && !defined(LIBXML_STATIC) */
+  #if !defined(LIBXML_STATIC)
     #define XMLPUBFUN __declspec(dllexport)
     #define XMLPUBVAR __declspec(dllexport)
   #else
